infer/environment.cc

```diff
@@ -30,6 +30,52 @@ bool typeTestReferencesVar(const InlinedVector<std::pair<cfg::LocalRef, core::Ty
 }
 } // namespace
 
+void TypeTestReverseIndex::addToIndex(cfg::LocalRef from, cfg::LocalRef to) {
+    auto &list = index[from];
+    // Maintain invariant: lists are sorted sets (no duplicate entries)
+    // lower_bound returns the first location that is >= to's ID, which is where `to` should be inserted if *it != to
+    // (uses binary search, O(log n))
+    auto it =
+        lower_bound(list.begin(), list.end(), to, [](const auto &a, const auto &b) -> bool { return a.id() < b.id(); });
+    if (it == list.end() || *it != to) {
+        list.insert(it, to);
+    }
+}
+
+const InlinedVector<cfg::LocalRef, 1> &TypeTestReverseIndex::get(cfg::LocalRef from) const {
+    auto it = index.find(from);
+    if (it == index.end()) {
+        return empty;
+    }
+    return it->second;
+}
+
+void TypeTestReverseIndex::replace(cfg::LocalRef from, InlinedVector<cfg::LocalRef, 1> &&list) {
+    index[from] = std::move(list);
+}
+
+void TypeTestReverseIndex::cloneFrom(const TypeTestReverseIndex &index) {
+    this->index = index.index;
```

Comment:
Why is it ok to not also copy `empty` here?

Reply:
`empty` is initialized as part of the default constructor and doesn't need to be consistent across objects. It's also `const` so I can't assign to it outside of the constructor.

Reply:
Would making `empty` static help...?

Reply:
Ah, a static empty would be great, then there'd only be one 👍